vyper/context/validation/local.py

```diff
@@ -350,9 +369,20 @@ def visit_Expr(self, node):
             raise StructureException("To call an event you must use the `log` statement", node)
 
         if isinstance(fn_type, ContractFunctionType):
-            if self.func.is_constant and not fn_type.is_constant:
+            if self.func.mutability is StateMutability.VIEW and fn_type.mutability not in (
+                StateMutability.VIEW,
+                StateMutability.PURE,
+            ):
+                raise ConstancyViolation(
+                    "Cannot call a mutating function from a view function", node
+                )
+
+            if (
+                self.func.mutability is StateMutability.PURE
+                and fn_type.mutability is not StateMutability.PURE
+            ):
                 raise ConstancyViolation(
-                    "Cannot call a non-constant function from a constant function", node
+                    "Cannot call a stateful function from a pure function", node
```

Comment:
Ahh..  "stateful" :point_up:  for previous comment.

Reply:
stateful does not mean state modifying though